src/structures/paging/recursive.rs

```diff
@@ -237,6 +276,224 @@ impl<'a> RecursivePageTable<'a> {
 
         inner(entry, next_table_page, allocator)
     }
+
+    /// This is a common implementation of `reclaim_page_tables`. See the docs for
+    /// `Mapper::reclaim_page_tables`.
+    unsafe fn reclaim_page_tables_common<D, R>(
+        &mut self,
+        range: R,
+        deallocator: &mut D,
+        mode: ReclaimPageTablesMode,
+    ) where
+        D: FrameDeallocator<Size4KiB>,
+        R: RangeBounds<Page<Size4KiB>>,
+    {
+        // Get the page ranges.
+        let start = match range.start_bound() {
+            Bound::Included(endpoint) => *endpoint,
+            Bound::Excluded(endpoint) => *endpoint + 1,
+            Bound::Unbounded => {
+                panic!("Attempt to reclaim page tables for unbounded region of memory.")
+            }
+        };
+        let end = match range.end_bound() {
+            Bound::Included(endpoint) => *endpoint + 1,
+            Bound::Excluded(endpoint) => *endpoint,
+            Bound::Unbounded => {
+                panic!("Attempt to reclaim page tables for unbounded region of memory.")
+            }
+        };
+
+        let p4 = &mut self.p4;
+
+        // Starting at the leaf page tables (p1), we reclaim page tables. Then we do p2 and p3.
+
+        // We can only reclaim a page table if all of its possible children are in the given range,
+        // so round `start` and `end` to exclude page tables that are only partial in range.
+        //
+        // That is, we round `start` up to the nearest 512 * Size4KiB boundary, and we round `end`
+        // down to the nearest boundary.
+
+        let p1_start = if start.p1_index() == u9::new(0) {
+            start
+        } else {
+            Page::from_page_table_indices(
+                start.p4_index(),
+                start.p3_index(),
+                start.p2_index(),
+                start.p1_index() + u9::new(1),
+            )
+        };
+
+        let p1_end = Page::from_page_table_indices(
+            start.p4_index(),
+            start.p3_index(),
+            start.p2_index(),
+            u9::new(0),
+        );
+
+        // Free all the page tables!
+        for page in p1_start..p1_end {
```

Comment:
Aren't you iterating over the same page table multiple times (same for the other page table levels)?

Reply:
Yes, but after you free the table the first time, it should short-circuit in subsequent iterations.

Reply:
Hmm, I think it would be less confusing to use 2MiB pages instead, because then each P1 table would occur only once.